### ISSUE TITLE
InAppBrowser issue with Apache Cordova 5.4.0/5.4.1 on Android

### ISSUE THREAD

**Issue by Jarvey** [nraboy/ng-cordova-oauth#174]
Hi, The release notes of the latest version is saying: Fix InAppBrowser issue with Apache Cordova 5.4.0.
I stille have a InAppBrowser error on Android with Apache Cordova 5.4.0 and 5.4.1: Oauth plugin is saying "Could not find InAppBrowser plugin" when trying to login via Oauth on Strava

**Comment by matheusrocha89** [comment]
@Jarvey I will test this, with me it was working with all major oauth providers.

**Comment by matheusrocha89** [comment]
@Jarvey working fine with me, are you sure you are using the latest version of our lib?

![strava](https://cloud.githubusercontent.com/assets/3718366/11685953/a827c57a-9e5b-11e5-95e6-4e3fc5ef2ca3.gif)

**Comment by matheusrocha89** [comment]
If you still have this issue with the new version (0.2.0) could you please paste your code here?

**Comment by nraboy** [comment]
My guess is that @Jarvey is trying to use ionic serve, ionic view, ionic live-reload or phonegap build, all of which are unsupported.

Curious to hear what is reported.

**Comment by Jarvey** [comment]
@nraboy , @matheusrocha89  I tested it on a android device (Samsung S4 and Samsung S4 mini) via Android studio. I had the same problem also on my iOS devices but after updating to 0.20 iOS app is working now. I'm using the latest cordova and Ionic.

```javascript
ng-cordova-oauth 0.20 code:
angular.module("oauth.providers", [
  "oauth.utils",
  'oauth.strava'])
  .factory("$cordovaOauth", cordovaOauth);

function cordovaOauth(
  $q, $http, $cordovaOauthUtility, $strava) {

  return {
    strava: $strava.signin,
  };
}

cordovaOauth.$inject = [
  "$q", '$http', "$cordovaOauthUtility",
  '$strava'
];


angular.module('oauth.strava', ['oauth.utils'])
  .factory('$strava', strava);

function strava($q, $http, $cordovaOauthUtility) {
  return { signin: oauthStrava };

  /*
  * Sign into the Strava service
  *
  * @param    string clientId
  * @param    string clientSecret
  * @param    array appScope
  * @param    object options
  * @return   promise
  */
  function oauthStrava(clientId, clientSecret, appScope, options) {
    var deferred = $q.defer();
    if(window.cordova) {
      if($cordovaOauthUtility.isInAppBrowserInstalled()) {
        var redirect_uri = "http://localhost/callback";
        if(options !== undefined) {
          if(options.hasOwnProperty("redirect_uri")) {
            redirect_uri = options.redirect_uri;
          }
        }
        var browserRef = window.cordova.InAppBrowser.open('https://www.strava.com/oauth/authorize?client_id=' + clientId + '&redirect_uri=' + redirect_uri + '&scope=' + appScope.join(",") + '&response_type=code&approval_prompt=force', '_blank', 'location=yes,clearsessioncache=yes,clearcache=yes');
        browserRef.addEventListener('loadstart', function(event) {
          if((event.url).indexOf(redirect_uri) === 0) {
            requestToken = (event.url).split("code=")[1];
            $http.defaults.headers.post['Content-Type'] = 'application/x-www-form-urlencoded';
            $http({method: "post", url: "https://www.strava.com/oauth/token", data: "client_id=" + clientId + "&client_secret=" + clientSecret + "&code=" + requestToken })
            .success(function(data) {
              deferred.resolve(data);
            })
            .error(function(data, status) {
              deferred.reject("Problem authenticating");
            })
            .finally(function() {
              setTimeout(function() {
                browserRef.close();
              }, 10);
            });
          }
        });
        browserRef.addEventListener('exit', function(event) {
          deferred.reject("The sign in flow was canceled");
        });
      } else {
        deferred.reject("Could not find InAppBrowser plugin");
      }
    } else {
      deferred.reject("Cannot authenticate via a web browser");
    }

    return deferred.promise;
  }
}

strava.$inject = ['$q', '$http', '$cordovaOauthUtility'];


/*
 * Cordova AngularJS Oauth
 *
 * Created by Nic Raboy
 * http://www.nraboy.com
 *
 *
 *
 * DESCRIPTION:
 *
 * Use Oauth sign in for various web services.
 *
 *
 * REQUIRES:
 *
 *    Apache Cordova 3.5+
 *    Apache InAppBrowser Plugin
 *    Apache Cordova Whitelist Plugin
 *
 *
 * SUPPORTS:
 *
 *    Strava
 */

angular.module("ngCordovaOauth", [
    "oauth.providers",
    "oauth.utils"
]);

angular.module("oauth.utils", [])
  .factory("$cordovaOauthUtility", cordovaOauthUtility);

function cordovaOauthUtility($q) {
  return {
    isInAppBrowserInstalled: isInAppBrowserInstalled,
    createSignature: createSignature,
    createNonce: createNonce,
    generateUrlParameters: generateUrlParameters,
    parseResponseParameters: parseResponseParameters,
    generateOauthParametersInstance: generateOauthParametersInstance
  };

  /*
   * Check to see if the mandatory InAppBrowser plugin is installed
   *
   * @param
   * @return   boolean
   */
  function isInAppBrowserInstalled() {
    var cordovaPluginList = cordova.require("cordova/plugin_list");
    var inAppBrowserNames = ["cordova-plugin-inappbrowser", "org.apache.cordova.inappbrowser"];

    if (Object.keys(cordovaPluginList.metadata).length === 0) {
      var formatedPluginList = cordovaPluginList.map(
        function(plugin) {
          return plugin.pluginId;
        });
      return inAppBrowserNames.some(function(name) {
        return formatedPluginList.indexOf(name) != -1 ? true : false;
      });
    } else {
      return inAppBrowserNames.some(function(name) {
        return cordovaPluginList.metadata.hasOwnProperty(name);
      });
    }
  }

  /*
   * Sign an Oauth 1.0 request
   *
   * @param    string method
   * @param    string endPoint
   * @param    object headerParameters
   * @param    object bodyParameters
   * @param    string secretKey
   * @param    string tokenSecret (optional)
   * @return   object
   */
  function createSignature(method, endPoint, headerParameters, bodyParameters, secretKey, tokenSecret) {
    if(typeof jsSHA !== "undefined") {
      var headerAndBodyParameters = angular.copy(headerParameters);
      var bodyParameterKeys = Object.keys(bodyParameters);

      for(var i = 0; i < bodyParameterKeys.length; i++) {
        headerAndBodyParameters[bodyParameterKeys[i]] = encodeURIComponent(bodyParameters[bodyParameterKeys[i]]);
      }

      var signatureBaseString = method + "&" + encodeURIComponent(endPoint) + "&";
      var headerAndBodyParameterKeys = (Object.keys(headerAndBodyParameters)).sort();

      for(i = 0; i < headerAndBodyParameterKeys.length; i++) {
        if(i == headerAndBodyParameterKeys.length - 1) {
          signatureBaseString += encodeURIComponent(headerAndBodyParameterKeys[i] + "=" + headerAndBodyParameters[headerAndBodyParameterKeys[i]]);
        } else {
          signatureBaseString += encodeURIComponent(headerAndBodyParameterKeys[i] + "=" + headerAndBodyParameters[headerAndBodyParameterKeys[i]] + "&");
        }
      }

      var oauthSignatureObject = new jsSHA(signatureBaseString, "TEXT");

      var encodedTokenSecret = '';
      if (tokenSecret) {
        encodedTokenSecret = encodeURIComponent(tokenSecret);
      }

      headerParameters.oauth_signature = encodeURIComponent(oauthSignatureObject.getHMAC(encodeURIComponent(secretKey) + "&" + encodedTokenSecret, "TEXT", "SHA-1", "B64"));
      var headerParameterKeys = Object.keys(headerParameters);
      var authorizationHeader = 'OAuth ';

      for(i = 0; i < headerParameterKeys.length; i++) {
        if(i == headerParameterKeys.length - 1) {
          authorizationHeader += headerParameterKeys[i] + '="' + headerParameters[headerParameterKeys[i]] + '"';
        } else {
          authorizationHeader += headerParameterKeys[i] + '="' + headerParameters[headerParameterKeys[i]] + '",';
        }
      }

      return { signature_base_string: signatureBaseString, authorization_header: authorizationHeader, signature: headerParameters.oauth_signature };
    } else {
      return "Missing jsSHA JavaScript library";
    }
  }

  /*
  * Create Random String Nonce
  *
  * @param    integer length
  * @return   string
  */
  function createNonce(length) {
    var text = "";
    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

    for(var i = 0; i < length; i++) {
      text += possible.charAt(Math.floor(Math.random() * possible.length));
    }

    return text;
  }

  function generateUrlParameters(parameters) {
    var sortedKeys = Object.keys(parameters);
    sortedKeys.sort();

    var params = "";
    var amp = "";

    for (var i = 0 ; i < sortedKeys.length; i++) {
      params += amp + sortedKeys[i] + "=" + parameters[sortedKeys[i]];
      amp = "&";
    }

    return params;
  }

  function parseResponseParameters(response) {
    if (response.split) {
      var parameters = response.split("&");
      var parameterMap = {};

      for(var i = 0; i < parameters.length; i++) {
          parameterMap[parameters[i].split("=")[0]] = parameters[i].split("=")[1];
      }

      return parameterMap;
    }
    else {
      return {};
    }
  }

  function generateOauthParametersInstance(consumerKey) {
    var nonceObj = new jsSHA(Math.round((new Date()).getTime() / 1000.0), "TEXT");
    var oauthObject = {
        oauth_consumer_key: consumerKey,
        oauth_nonce: nonceObj.getHash("SHA-1", "HEX"),
        oauth_signature_method: "HMAC-SHA1",
        oauth_timestamp: Math.round((new Date()).getTime() / 1000.0),
        oauth_version: "1.0"
    };
    return oauthObject;
  }
}
cordovaOauthUtility.$inject = ['$q']; 
```

My Strava login button:

```javascript
$scope.login = function() {
		
        $cordovaOauth.strava("clientId", "string clientSecret", ["view_private"]).then(function(result) {
          	
            $localStorage.stravaInfo = {
	            "accessToken" : result.access_token,
	            "athleteId" : result.athlete.id,
	            "athleteFirstname" : result.athlete.firstname,
	            "athleteLastname" : result.athlete.lastname,
	            "athleteProfile" : result.athlete.profile,
	            "athleteClubs" : result.athlete.clubs
            };
            
        	if ($localStorage.stravaInfo.athleteClubs.length == 0) {
            	//go to this page when there are no clubs available
				$state.go("noClubs");
            
            } else {
	            //go to profile page to compleet setup
	            $state.go("profile");
	        }	
	                    
        }, function(error) {
	        alert(error);
            $state.go("login");
        });
    };

**Comment by nraboy** [comment]
So just to be clear, everything is good now?

**Comment by Jarvey** [comment]
@nraboy On iOS yes, Android no still error: Could not find InAppBrowser plugin.
I have installed cordova-plugin-inappbrowser 1.1.0

**Comment by nraboy** [comment]
What happens if you wrap the call inside an `$ionicPlatform.ready`?  I'm wondering if there is a race condition happening and the cordova plugins haven't finished loading for Android.  This would be the first I hear of it if thats the case.

**Comment by Jarvey** [comment]
@nraboy  I wrapped the code below in ```javascript $ionicPlatform.ready ``` but with no succes, still error: Could not find InAppBrowser plugin

```javascript
$scope.login = function() {

        $cordovaOauth.strava("clientId", "string clientSecret", ["view_private"]).then(function(result) {

            $localStorage.stravaInfo = {
                "accessToken" : result.access_token,
                "athleteId" : result.athlete.id,
                "athleteFirstname" : result.athlete.firstname,
                "athleteLastname" : result.athlete.lastname,
                "athleteProfile" : result.athlete.profile,
                "athleteClubs" : result.athlete.clubs
            };

            if ($localStorage.stravaInfo.athleteClubs.length == 0) {
                //go to this page when there are no clubs available
                $state.go("noClubs");

            } else {
                //go to profile page to compleet setup
                $state.go("profile");
            }   

        }, function(error) {
            alert(error);
            $state.go("login");
        });
    };

**Comment by nraboy** [comment]
Can you confirm that you are using the latest Android Apache Cordova build?  This is not to be confused with the latest Apache Cordova build as platform builds are updated independently from the CLI.

```
cordova platform check
cordova platform update android
```

**Comment by Jarvey** [comment]
@nraboy When I run “cordova platform check” the response is:
iOS project created with cordova-ios@3.9.2
add to body class: platform-ios
android @ 5.0.0; current did not install, and thus its version cannot be determined

Then when I run “cordova platform update android” the response is:
Android project updated with cordova-android@4.1.1
Running command: /path/to/app…
add to body class: platform-android

It looks like it did not update to the latest version so I also tried running “cordova platform update android@5.0.0”, response:
Android project updated with cordova-android@5.0.0
Running command: /path/to/app…
add to body class: platform-android

Then I run “cordova platform check” again with the same response as the first time: 
iOS project created with cordova-ios@3.9.2
add to body class: platform-ios
android @ 5.0.0; current did not install, and thus its version cannot be determined

When I run "cordova -v", response is: 5.4.1
When I run "ionic -v," response is: 1.7.11
When I run "cordova platform ls" response is: Installed platforms: android 5.0.0, ios 3.9.2

**Comment by matheusrocha89** [comment]
@Jarvey Do this.
```
npm install -g cordova #To update cordova
ionic platform rm android
ionic platform add android
ionic run android 
```

**Comment by Jarvey** [comment]
@matheusrocha89 after running your commands I still get the error in the app.
When I ran the command "ionic platform add android" terminal was saying this:
>
Adding android project...
Creating Cordova project for the Android platform:
	Path: platforms/android
	Package: com.appname
	Name: appname
	Activity: MainActivity
	Android target: android-23
Android project created with cordova-android@5.0.0
Running command: /Users/username/folder/folder/folder/appname/hooks/after_prepare/010_add_platform_class.js /Users/username/folder/folder/folder/AppName
add to body class: platform-android
Installing "cordova-plugin-3dtouch" for android
Installing "cordova-plugin-camera" for android
Installing "cordova-plugin-device" for android
Installing "cordova-plugin-dialogs" for android
Installing "cordova-plugin-file" for android
>
The Android Persistent storage location now defaults to "Internal". Please check this plugins README to see if you application needs any changes in its config.xml.
>      
If this is a new application no changes are required.
>    
If this is an update to an existing application that did not specify an "AndroidPersistentFileLocation" you may need to add:
 >     
      "<preference name="AndroidPersistentFileLocation" value="Compatibility" />"
>   
to config.xml in order for the application to find previously stored files.
>    
Installing "cordova-plugin-file-transfer" for android
Dependent plugin "cordova-plugin-file" already installed on android.
Installing "cordova-plugin-globalization" for android
Installing "cordova-plugin-inappbrowser" for android
Installing "cordova-plugin-splashscreen" for android
Installing "cordova-plugin-whitelist" for android
>
This plugin is only applicable for versions of cordova-android greater than 4.0. If you have a previous platform version, you do *not* need this plugin since the whitelist will be built in.
>
Installing "de.appplant.cordova.plugin.email-composer" for android
Installing "ionic-plugin-keyboard" for android
Installing "phonegap-plugin-push" for android

After running the commands I also run "ionic platform check" with the same response as before:
>
iOS project created with cordova-ios@3.9.2
add to body class: platform-ios
android @ 5.0.0; current did not install, and thus its version cannot be determined

**Comment by matheusrocha89** [comment]
@Jarvey the red flags are no problems, don't worry, but you still have the error on Android? It's strange cause on the video I made I tested on Android and it's working fine. I think there is something wrong on your android enviroment, try to remove and add the inappbrowser to see if this make your app works. Just to know, before the update to 0.2.0 your app was working on android? If the problem persist I want to make a call with you to understand what is happening.

**Comment by Jarvey** [comment]
@matheusrocha89 I figured it out, after trying different things the last couple of days, a java SDK upgrade from version 7 to 8 did it. OAuth login in my app is now working again 😀. Thank you for you're great help!!

**Comment by nraboy** [comment]
Can this be resolved now?

**Comment by Jarvey** [comment]
@nraboy matheusrocha89 is write I think there is something wrong with my android environment. I updated my java sdk version from 7 to 8 and Oauth in de android app was working again on android. unfortunately yesterday when I run "ionic platform update android" it was broken again. Today when I removed the android platform and added the android platform again it oauth in android was working again. So I have no idea what is causing it but for now I'm fine. Thanks for you help!!
